new(modern_bpf): add support for `notify` syscalls

Andreagit97 opened the pull request:
**What type of PR is this?**

/kind feature

**Any specific area of the project related to this PR?**

/area driver-modern-bpf

/area libpman

/area tests

**Does this PR require a change in the driver versions?**

**What this PR does / why we need it**:

This PR is part of a series https://github.com/falcosecurity/libs/issues/513, the final aim is to support the most important syscalls also in the new probe. This PR introduces:

* `eventfd`
* `eventfd2`
* `inotify_init`
* `inotify_init1`
* `timerfd_create`
* `userfaultfd`
* `signalfd`
* `signalfd4`

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
new(modern_bpf): add support for `notify` syscalls 
```

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/516#" title="Author self-approved">Andreagit97</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

Andreagit97 commented on driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/eventfd.bpf.c:
```diff
@@ -0,0 +1,74 @@
+	ringbuf__store_u64(&ringbuf, (u64)initval);
+
+	/* Parameter 2: flags (type: PT_FLAGS32) */
+	/// TODO: Right now we don't catch any flag.
+	u32 flags = 0;
```
See issue https://github.com/falcosecurity/libs/issues/515

Andreagit97 commented on driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/signalfd.bpf.c:
```diff
@@ -0,0 +1,78 @@
+	ringbuf__store_s64(&ringbuf, (s64)fd);
+
+	/* Parameter 2: mask (type: PT_UINT32) */
+	/* Like in the old probe we send `0`. */
```
See issue https://github.com/falcosecurity/libs/issues/515

Andreagit97 commented on driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/timerfd_create.bpf.c:
```diff
@@ -0,0 +1,71 @@
+	/*=============================== COLLECT PARAMETERS  ===========================*/
+
+	/* Parameter 1: clockid (type: PT_UINT8) */
+	/* Like in the old probe we send `0` */
```
See issue https://github.com/falcosecurity/libs/issues/515

hbrueckner commented on test/modern_bpf/test_suites/syscall_enter_suite/eventfd_e.cpp:
```diff
@@ -0,0 +1,46 @@
+
+	uint32_t initval = 3;
+	int flags = 0;
+	int32_t fd = syscall(__NR_eventfd, initval, flags);
```
The `eventfd` and `eventfd2` syscalls seem to be interesting. Actually based on the [man page of eventfd](https://man7.org/linux/man-pages/man2/eventfd.2.html), it states

> There are two underlying Linux system calls: eventfd() and the more recent eventfd2().  The former system call does not implement a flags argument.  The latter system call implements the flags values described above.  The glibc wrapper function will use eventfd2() where it is available.

The Linux kernel source code seems to confirm that [here](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/fs/eventfd.c#n454).

So for the `eventfd`, the flags argument should not be necessary... only for the `eventfd2` syscall.

Was the scap event then actually implemented with `eventfd2` in mind?

Andreagit97 replied:
Yeah, that is the same case of `inotify_init` and `inotify_init1`, `inotify_init` doesn't have the flag but we catch it anyway. The solution could be to create 2 separate events as I have already tracked it in https://github.com/falcosecurity/libs/issues/515 :thinking:

hbrueckner replied:
Ok.. will look at them as well... but still I wonder passing flags to the syscall even if the syscall does not use it. And agree, I would rather split those events into 2 distinct ones.

Andreagit97 replied:
During this refactor we need to understand which params we need because I agree with you that some of these params are not used at all!!

hbrueckner commented on test/modern_bpf/test_suites/syscall_enter_suite/eventfd2_e.cpp:
```diff
@@ -0,0 +1,52 @@
+	evt_test->assert_numeric_param(1, (uint64_t)initval);
+
+	/* Parameter 2: flags (type: PT_FLAGS32) */
+	/* Right now we don't catch any flag, so we expect `0` as a second parameter. */
```
Being tracked here: #515

Andreagit97 replied:
great catch! Done, thank you!

hbrueckner commented on test/modern_bpf/test_suites/syscall_enter_suite/inotify_init_e.cpp:
```diff
@@ -0,0 +1,42 @@
+	/*=============================== ASSERT PARAMETERS  ===========================*/
+
+	/* Parameter 1: flags (type: PT_FLAGS8) */
+	evt_test->assert_numeric_param(1, (uint8_t)flags);
```
hmm... so if inotify_init(void) does not process the flags, then the BPF program just passes the first argument simply thru? How does it look like for a simply `inotify_init()` call where the BPF program then reports the (whatever) content the register for the first argument contains? If so, could that be a security issue/information leak?

Andreagit97 replied:
Not sure if this is an `information leak` but it could be... In my tests, it always returns `0`, but obviously we are not sure about the content, we must fix it!

hbrueckner replied:
> Not sure if this is an `information leak` but it could be...

It could be.... and depending on the architecture ELF ABI, argument registers might become call-clobbered/cleared etc.  However, being on the safe side is better.

hbrueckner replied:
Thanks for [flags removal in `eventfd`](https://github.com/falcosecurity/libs/pull/516/commits/ad4a3d9a42b09b1f6e0ce5425b12ba5ca16cae2a) which also had the potential for a leak.

hbrueckner commented on driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/timerfd_create.bpf.c:
```diff
@@ -0,0 +1,71 @@
+
+	/* Parameter 2: flags (type: PT_FLAGS8) */
+	/* Like in the old probe we send `0` */
+	ringbuf__store_u8(&ringbuf, 0);
```
I guess for both, helpers for to map to scap events are also necessary then.

hbrueckner commented on test/modern_bpf/test_suites/syscall_enter_suite/signalfd4_e.cpp:
```diff
@@ -0,0 +1,59 @@
+	int32_t mock_fd = -1;
+	sigset_t mask = {0};
+	int flags = 7;
+	assert_syscall_state(SYSCALL_FAILURE, "signalfd4", syscall(__NR_signalfd4, mock_fd, &mask, flags));
```
This also relates to the `signalfd` BPF program. Looks like there is a similar pattern like described [here](https://github.com/falcosecurity/libs/pull/516/commits/149c181dbf735d61f1ce33a7b0462786ddbb13a7#r935482912).

According to the man page, the system call includes an additional argument, `size_t sizemask` which is transparently handled by the glibc wrapper. For `signalfd4` it is the third argument (where `flags` is passed). I would assume that for calling `signalfd` and `signalfd4` directly, this parameter needs to be added here as well. WDYT?

(I think for now, this is  a blind spot because the BPF programs do not catch those arguments)

For reference, [syscall definition in the Linux kernel](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/fs/signalfd.c#n302).

Andreagit97 replied:
This is really a great catch thank you for your accuracy!! :heart_eyes:

Andreagit97 commented:
I have added all your comments in the tracking issue, there is a lot of work to do on that :/

hbrueckner commented:
Hi @Andreagit97 

> @hbrueckner I have added all your comments on the tracking issue, there is a lot of work to do on that :/

Thanks for adding them all! and sorry for all the new todos

Andreagit97 commented:
> Hi @Andreagit97
> 
> > @hbrueckner I have added all your comments on the tracking issue, there is a lot of work to do on that :/
> 
> Thanks for adding them all! and sorry for all the new todos

I should have addressed all the comments and I have tracked all the problems in our issue https://github.com/falcosecurity/libs/issues/515. @hbrueckner thank you for the review as always! Just one note, to temporarily address the possible information leakage in [`inotify_init`](https://github.com/falcosecurity/libs/pull/516#discussion_r935500708) I have removed the `flag` collection in the `PPME_SYSCALL_INOTIFY_INIT_E` event. This was quite useless even because we didn't convert the flags in the scap format... To fix this stuff we will introduce 2 different events, one for the `inotify_init` and the other for `inotify_init1`

hbrueckner commented:
Hi @Andreagit97

> I should have addressed all the comments and I have tracked all the problems in our issue #515. @hbrueckner thank you for the review as always! Just one note, to temporarily address the possible information leakage in [`inotify_init`](https://github.com/falcosecurity/libs/pull/516#discussion_r935500708) I have removed the `flag` collection in the `PPME_SYSCALL_INOTIFY_INIT_E` event. This was quite useless even because we didn't convert the flags in the scap format... To fix this stuff we will introduce 2 different events, one for the `inotify_init` and the other for `inotify_init1`

Thank you very much! LGTM and I will give them a try on s390 later this week and share test results.

hbrueckner commented:
Hi @Andreagit97

> Thank you very much! LGTM and I will give them a try on s390 later this week and share test results.

Here we go, all tests have passed as well:
````
[ RUN      ] SyscallExit.eventfd2X
[       OK ] SyscallExit.eventfd2X (0 ms)
[ RUN      ] SyscallExit.eventfdX
[       OK ] SyscallExit.eventfdX (0 ms)
[ RUN      ] SyscallExit.inotify_init1X
[       OK ] SyscallExit.inotify_init1X (0 ms)
[ RUN      ] SyscallExit.inotify_initX
[       OK ] SyscallExit.inotify_initX (0 ms)
[ RUN      ] SyscallExit.signalfd4X
[       OK ] SyscallExit.signalfd4X (0 ms)
[ RUN      ] SyscallExit.signalfdX
[       OK ] SyscallExit.signalfdX (0 ms)
[ RUN      ] SyscallExit.timerfd_createX
[       OK ] SyscallExit.timerfd_createX (0 ms)
[ RUN      ] SyscallExit.userfaultfdX
[       OK ] SyscallExit.userfaultfdX (0 ms)

[ RUN      ] SyscallEnter.eventfd2E
[       OK ] SyscallEnter.eventfd2E (0 ms)
[ RUN      ] SyscallEnter.eventfdE
[       OK ] SyscallEnter.eventfdE (0 ms)
[ RUN      ] SyscallEnter.inotify_init1E
[       OK ] SyscallEnter.inotify_init1E (0 ms)
[ RUN      ] SyscallEnter.inotify_initE
[       OK ] SyscallEnter.inotify_initE (0 ms)
[ RUN      ] SyscallEnter.signalfd4E
[       OK ] SyscallEnter.signalfd4E (0 ms)
[ RUN      ] SyscallEnter.signalfdE
[       OK ] SyscallEnter.signalfdE (0 ms)
[ RUN      ] SyscallEnter.timerfd_createE
[       OK ] SyscallEnter.timerfd_createE (0 ms)
[ RUN      ] SyscallEnter.userfaultfdE
[       OK ] SyscallEnter.userfaultfdE (0 ms)
````

FedeDP reviewed (approved):
/approve

poiana commented:
LGTM label has been added.  <details>Git tree hash: 437013d555e7d31d1aeb1e35c20c07502c8a3803</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/516#" title="Author self-approved">Andreagit97</a>*, *<a href="https://github.com/falcosecurity/libs/pull/516#pullrequestreview-1060263098" title="Approved">FedeDP</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [Andreagit97,FedeDP]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
